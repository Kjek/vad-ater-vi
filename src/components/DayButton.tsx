import { useGlobalState } from '@hook/useGlobalState';
import { useCallback } from 'react';
import type { AllWeekButtonProps } from './AllWeekButton';

interface DayButtonProps extends AllWeekButtonProps {
  title: string;
}

const DayButton = (props: DayButtonProps) => {
  const { state, dispatch } = useGlobalState();
  const { title, setAllSelected } = props;
  const isSelected = state.daysSelected.get(title);

  const onClick = useCallback(
    () => {
      dispatch({
        type: 'selected',
        payload: {
          day: title,
        },
      }),
        setAllSelected(false);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  const selectedClasses = isSelected
    ? 'z-10 outline-none ring-2 ring-gray-700 duration-75 bg-white dark:ring-gray-400 dark:bg-gray-800'
    : '';

  return (
    <>
      <input
        title='Tryck för att visa den här dagens lunch i listan nedan'
        className={`cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 transition-all duration-500 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white ${selectedClasses}`}
        type='button'
        value={props.title}
        onClick={onClick}
      />
    </>
  );
};

export default DayButton;
