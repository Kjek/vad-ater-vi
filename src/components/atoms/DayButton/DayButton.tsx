import { useGlobalState } from '@hook/useGlobalState';
import { cn } from '@util/cn';
import type { Dispatch, InputHTMLAttributes, SetStateAction } from 'react';
import { useCallback } from 'react';

interface DayButtonProps extends InputHTMLAttributes<HTMLInputElement> {
  title: string;
  setAllSelected: Dispatch<SetStateAction<boolean>>;
}

const DayButton = ({ title, setAllSelected, ...props }: DayButtonProps) => {
  const { state, dispatch } = useGlobalState();
  const isSelected = state.daysSelected.get(title);

  const onClick = useCallback(
    () => {
      dispatch({
        type: 'selected',
        payload: {
          day: title,
        },
      }),
        setAllSelected(false);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  const selectedClasses = isSelected
    ? 'z-10 outline-none ring-2 ring-gray-700 bg-white dark:ring-gray-400 dark:bg-gray-800'
    : '';

  return (
    <>
      <input
        {...props}
        title='Tryck för att visa den här dagens lunch i listan nedan'
        className={cn(
          'cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 transition duration-500 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white',
          selectedClasses
        )}
        type='button'
        value={title}
        onClick={onClick}
      />
    </>
  );
};

export default DayButton;
