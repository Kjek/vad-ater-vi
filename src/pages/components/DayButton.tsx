import { useCallback } from 'react';
import { useGlobalState } from '~/hooks/useGlobalState';

interface DayButtonProps {
  title: string;
}

const DayButton = (props: DayButtonProps) => {
  const { state, dispatch } = useGlobalState();
  const { title } = props;
  const isSelected = state.daysSelected.get(title);

  const onClick = useCallback(
    () =>
      dispatch({
        type: state.isMultiSelect ? 'multi-select' : 'single-select',
        payload: {
          isSelected: !isSelected,
          day: title,
        },
      }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [state]
  );

  const selectedClasses = isSelected
    ? 'z-10 outline-none ring-4 ring-gray-200 duration-75 dark:ring-gray-700'
    : '';

  return (
    <>
      <input
        className={`mr-2 mb-2 cursor-pointer rounded-lg border border-gray-200 bg-white py-2.5 px-5 text-sm font-medium text-gray-900 transition-all duration-500 hover:bg-gray-100 hover:text-blue-700 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white ${selectedClasses}`}
        type='button'
        value={props.title}
        onClick={onClick}
      />
    </>
  );
};

export default DayButton;
