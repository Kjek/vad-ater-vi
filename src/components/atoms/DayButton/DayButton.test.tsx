import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import DayButton from './DayButton';

describe(DayButton, () => {
  let isAllSelected = true;
  const setAllSelected = vi.fn().mockImplementation((newValue: boolean) => {
    isAllSelected = newValue;
  });

  beforeEach(() => {
    vi.clearAllMocks();
    isAllSelected = false;
  });

  it('renders correctly', () => {
    const { getByTestId } = render(
      <DayButton
        data-testid={'day.button'}
        setAllSelected={setAllSelected}
        title='Måndag'
      />
    );
    const dayButton = getByTestId('day.button') as HTMLInputElement;
    expect(dayButton).toBeInTheDocument();
    expect(dayButton).toBeVisible();
  });

  it('displays correctly', () => {
    const { getByTestId } = render(
      <DayButton
        data-testid={'day.button'}
        setAllSelected={setAllSelected}
        title='Måndag'
      />
    );
    const dayButton = getByTestId('day.button') as HTMLSpanElement;
    expect(dayButton).toBeVisible();
    expect(dayButton).toHaveValue('Måndag');
  });

  it('changing state correctly', () => {
    const { getByTestId } = render(
      <DayButton
        data-testid={'day.button'}
        setAllSelected={setAllSelected}
        title='Måndag'
      />
    );
    const dayButton = getByTestId('day.button') as HTMLSpanElement;
    expect(dayButton).toBeInTheDocument();
    expect(dayButton).toBeVisible();

    fireEvent.click(dayButton);

    expect(setAllSelected).toBeCalledWith(false);
    expect(isAllSelected).toBeFalsy();
  });

  it('matches snapshot', () => {
    const defaultRender = render(
      <DayButton
        data-testid={'day.button'}
        setAllSelected={setAllSelected}
        title='Måndag'
      />
    );
    expect(defaultRender).toMatchSnapshot();
  });
});
