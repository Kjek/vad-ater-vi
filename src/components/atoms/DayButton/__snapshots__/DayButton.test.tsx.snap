// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DayButton > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <input
        class="cursor-pointer rounded-lg border border-gray-300 px-3.5 py-2 text-sm font-medium text-gray-800 transition duration-500 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white md:px-5 md:py-2.5"
        data-testid="day.button"
        title="Tryck för att visa den här dagens lunch i listan nedan"
        type="button"
        value="Måndag"
      />
    </div>
  </body>,
  "container": <div>
    <input
      class="cursor-pointer rounded-lg border border-gray-300 px-3.5 py-2 text-sm font-medium text-gray-800 transition duration-500 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white md:px-5 md:py-2.5"
      data-testid="day.button"
      title="Tryck för att visa den här dagens lunch i listan nedan"
      type="button"
      value="Måndag"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
